File: java/leetcode/backtracking/permutations_ii.java
/**
 * 47. Permutations II
 * Given a collection of numbers that might contain duplicates, return all possible unique permutations.

Example:

Input: [1,1,2]
Output:
[
  [1,1,2],
  [1,2,1],
  [2,1,1]
]
 */

import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;


public class permutations_ii {

    public List<List<Integer>> permuteUnique(int[] nums) {
        List<List<Integer>> permutes = new ArrayList<>();
        List<Integer> permuteList = new ArrayList<>();
        Arrays.sort(nums);  // sort
        boolean[] hasVisited = new boolean[nums.length];
        backtracking(permuteList, permutes, hasVisited, nums);
        return permutes;
    }

    private void backtracking(List<Integer> permuteList, List<List<Integer>> permutes, 
        boolean[] visited, final int[] nums) {
        
        if (permuteList.size() == nums.length) {
            permutes.add(new ArrayList<>(permuteList));
            return;
        }

        for (int i = 0; i < visited.length; i++) {
            if (i != 0 && nums[i] == nums[i-1] && !visited[i-1]) {
                continue;  // avoid adding repeated elements
            }
            if (visited[i]) {
                continue;
            }
            visited[i] = true;
            permuteList.add(nums[i]);
            backtracking(permuteList, permutes, visited, nums);
            permuteList.remove(permuteList.size() - 1);
            visited[i] = false;
        }
    }

}

File: golang/leetcode/backtracking/permutations_ii.go
package backtracking

import "sort"

// 47. Permutations II

func permuteUnique(nums []int) [][]int {
	permutes := make([][]int, 0)
	temp := make([]int, 0)
	sort.Ints(nums)
	visited := make([]bool, len(nums))
	permuteTrack2(temp, &permutes, visited, nums)
	return permutes
}

func permuteTrack2(temp []int, permutes *[][]int, visited []bool, nums []int) {
	if len(temp) == len(nums) {
		combo := make([]int, len(temp))
		copy(combo, temp)
		*permutes = append(*permutes, combo)
		return
	}
	for i := 0; i < len(visited); i++ {
		if i != 0 && nums[i] == nums[i-1] && !visited[i-1] {
			continue // avoid adding repeated elements
		}
		if visited[i] {
			continue
		}
		visited[i] = true
		temp = append(temp, nums[i])
		permuteTrack2(temp, permutes, visited, nums)
		temp = temp[:len(temp)-1]
		visited[i] = false
	}
}
